File: uni-idea-plugin/src/main/java/com/intellij/ide/projectView/impl/SpeedSearchFiles.java
package com.intellij.ide.projectView.impl;

import com.intellij.ide.projectView.impl.nodes.PsiDirectoryNode;
import com.intellij.openapi.util.text.StringUtil;
import com.intellij.ui.TreeSpeedSearch;

import javax.swing.*;
import javax.swing.tree.DefaultMutableTreeNode;
import javax.swing.tree.TreePath;
import java.util.StringTokenizer;

public final class SpeedSearchFiles extends TreeSpeedSearch {
  public SpeedSearchFiles(JTree tree) {
    super(tree);
  }

  @Override
  protected boolean isMatchingElement(Object element, String pattern) {
    Object userObject = ((DefaultMutableTreeNode) ((TreePath) element).getLastPathComponent()).getUserObject();
    if (userObject instanceof PsiDirectoryNode) {
      String str = getElementText(element);
      if (str == null) return false;
      str = StringUtil.toLowerCase(str);
      if (pattern.indexOf('.') >= 0) {
        return compare(str, pattern);
      }
      StringTokenizer tokenizer = new StringTokenizer(str, ".");
      while (tokenizer.hasMoreTokens()) {
        String token = tokenizer.nextToken();
        if (compare(token, pattern)) {
          return true;
        }
      }
      return false;
    } else {
      return super.isMatchingElement(element, pattern);
    }
  }
}

File: uni-idea-plugin/src/main/java/com/intellij/ide/projectView/impl/SpeedSearchFiles.kt
package com.intellij.ide.projectView.impl

import com.intellij.ide.projectView.impl.nodes.PsiDirectoryNode
import com.intellij.openapi.util.text.StringUtil
import java.util.*
import javax.swing.JTree
import javax.swing.tree.DefaultMutableTreeNode
import javax.swing.tree.TreePath

class SpeedSearchFiles(tree: JTree?) : TreeSpeedSearch2(tree) {
  override fun isMatchingElement(element: Any, pattern: String): Boolean {
    val userObject = ((element as TreePath).lastPathComponent as DefaultMutableTreeNode).userObject
    return if (userObject is PsiDirectoryNode) {
      var str: String? = getElementText(element) ?: return false
      str = StringUtil.toLowerCase(str)
      if (pattern.indexOf('.') >= 0) {
        return compare(str, pattern)
      }
      val tokenizer = StringTokenizer(str, ".")
      while (tokenizer.hasMoreTokens()) {
        val token = tokenizer.nextToken()
        if (compare(token, pattern)) {
          return true
        }
      }
      false
    } else {
      super.isMatchingElement(element, pattern)
    }
  }
}